vgaidarji/github-android-app
# Build contributions calendar component

### Context
Need to build the contributions UI component. 
<img width="893" alt="image" src="https://github.com/vgaidarji/droidhub-android-app/assets/3036347/84b4be16-bffe-41ff-adec-8708b32d0432">

Ideally have the ability to switch between years and then show only 1 single month contributions to fit the screen and then be able to navigate between months for given year: 
```
                  <contributions year>
<previous month> <contributions component> <next month>
```
or if it's easier and works nicely as well have contributions grid be rendered for given year and add horizontal scroll as it works on GitHub.

- https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-github-profile/managing-contribution-settings-on-your-profile/viewing-contributions-on-your-profile
- https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-github-profile/managing-contribution-settings-on-your-profile/why-are-my-contributions-not-showing-up-on-my-profile

Contributions are available only GraphQL API version again. Use GraphQL explorer to test the responses https://docs.github.com/en/graphql/overview/explorer

`GraphQL API query to fetch contributions`
```graphql
query { 
  user(login: "vgaidarji") {
    email
    createdAt
    contributionsCollection(from: "2020-01-01T00:00:00Z", to: "2020-12-31T23:59:59Z") {
      contributionCalendar {
        totalContributions
        weeks {
          contributionDays {
            weekday
            date 
            contributionCount 
            color
          }
        }
        months  {
          name
            year
            firstDay 
          totalWeeks 
          
        }
      }
    }
  }
} 
```

`Contributions response`
```json
{
  "data": {
    "user": {
      "email": "veaceslav.gaidarji@gmail.com",
      "createdAt": "2012-12-13T18:53:49Z",
      "contributionsCollection": {
        "contributionCalendar": {
          "totalContributions": 308,
          "weeks": [
            {
              "contributionDays": [
                {
                  "weekday": 3,
                  "date": "2020-01-01T00:00:00.000+00:00",
                  "contributionCount": 0,
                  "color": "#ebedf0"
                },
                {
                  "weekday": 4,
                  "date": "2020-01-02T00:00:00.000+00:00",
                  "contributionCount": 1,
                  "color": "#9be9a8"
                },
                {
                  "weekday": 5,
                  "date": "2020-01-03T00:00:00.000+00:00",
                  "contributionCount": 1,
                  "color": "#9be9a8"
                },
                {
                  "weekday": 6,
                  "date": "2020-01-04T00:00:00.000+00:00",
                  "contributionCount": 0,
                  "color": "#ebedf0"
                }
              ]
            },
...
```
and response contains contributions as well as color coding based on number of contributions (so I don't need to come up with mapping the # of contributions to the color and can use it straight away from response which is cool)

### What has been done
- Implemented contributions calendar UI component. Contributions data is only available in GraphQL API.
- Used compose preview parametrization to provide contributions response and test contributions calendar rendering with real data

[Screen_recording_20240712_111511.webm](https://github.com/user-attachments/assets/a0065cc7-bf46-4389-9157-27638d70db25)


### How to test
- Use previews and confirm that contributions calendar renders correctly for preview mock data
- Test contributions for different years on device and confirm they are properly shown and match with GitHub contributions on the GitHub website
